agent-shell.el

```diff
@@ -389,6 +455,88 @@ Returns an empty string if no icon should be displayed."
+Note: The returned state should be treated as read-only by extensions."
+  (with-current-buffer (or buffer (current-buffer))
+    (when (and (boundp 'agent-shell--state)
+               (derived-mode-p 'agent-shell-mode))
+      agent-shell--state)))
+
+(defun agent-shell-get-client (&optional buffer)
+  "Get ACP client from BUFFER's agent shell state.
+BUFFER defaults to current buffer.
+Returns nil if buffer is not an agent-shell buffer."
+  (map-elt (agent-shell-get-state buffer) :client))
+
+(defun agent-shell-tool-call-get (tool-call-id &optional buffer)
+  "Get tool call data for TOOL-CALL-ID from BUFFER's agent shell state.
+BUFFER defaults to current buffer.
+Returns the tool call alist or nil if not found."
+  (when-let ((state (agent-shell-get-state buffer)))
+    (map-nested-elt state (list :tool-calls tool-call-id))))
+
+(defun agent-shell-tool-call-put (tool-call-id data &optional buffer)
+  "Store tool call DATA for TOOL-CALL-ID in BUFFER's agent shell state.
+BUFFER defaults to current buffer.
+DATA should be an alist that will be merged with existing data.
+Returns non-nil on success."
+  (when-let ((state (agent-shell-get-state buffer)))
+    (agent-shell--save-tool-call state tool-call-id data)
+    t))
+
+(defun agent-shell-resolve-path (path)
+  "Resolve PATH using configured path resolver.
+This applies any path transformations configured via
+`agent-shell-path-resolver-function'.
+
+Extensions should use this instead of `agent-shell--resolve-path'
+to ensure consistent path handling across the system."
+  (agent-shell--resolve-path path))
+
+(defun agent-shell-send-permission-response (request-id option-id &optional buffer)
+  "Send permission response for REQUEST-ID with OPTION-ID.
+BUFFER defaults to current buffer.
+Returns non-nil on success.
+
+This function is intended for use by extensions implementing custom
+permission queueing via `agent-shell-permission-request-functions'.
+It sends the ACP response and cleans up the permission dialog UI.
+
+Extensions that intercept permissions are responsible for calling this
+function to send responses at the appropriate time."
+  (when-let ((state (agent-shell-get-state buffer))
+             (client (map-elt state :client)))
+    (with-current-buffer (or buffer (current-buffer))
+      ;; Find the tool-call-id for this request
+      (when-let ((tool-call-entry
+                  (seq-find (lambda (entry)
+                              (equal (map-elt (cdr entry) :permission-request-id)
+                                     request-id))
+                            (map-elt state :tool-calls))))
+        (let ((tool-call-id (car tool-call-entry)))
+          (agent-shell--send-permission-response
+           :client client
+           :request-id request-id
+           :option-id option-id
+           :state state
+           :tool-call-id tool-call-id)
+          t)))))
+
+(defun agent-shell-delete-dialog-block (block-id &optional buffer)
```

Comment:
cl-defun agent-shell-delete-dialog-block (&key block-id buffer)

ps. I've been considering renaming `dialog` in most places to `section`. Just a heads-up. If you're keen to adopt `section` now, that works for me. Otherwise I can ping you when I rename it or send a PR to your package.